packages/strapi/lib/middlewares/cors/index.js

```diff
@@ -32,8 +32,16 @@ module.exports = strapi => {
 
       strapi.app.use(
         cors({
-          origin: function(ctx) {
-            const whitelist = Array.isArray(origin) ? origin : origin.split(/\s*,\s*/);
+          origin: async function(ctx) {
+            let originList;
+
+            if (typeof origin === 'function') {
+              originList = await origin(ctx);
+            } else {
+              originList = origin;
+            }
+
+            const whitelist = Array.isArray(originList) ? originList : originList.split(/\s*,\s*/);
```

Comment:
You can simplify with:

```javascript
            let whitelist;

            if (typeof origin === 'function') {
              whitelist = await origin(ctx);
            } else {
              whitelist = Array.isArray(origin) ? origin : origin.split(/\s*,\s*/);
            }
```

Reply:
or 

```javascript
            let whitelist;

            if (typeof origin === 'function') {
              whitelist = await origin(ctx);
            } else if (typeof origin === 'string') {
              whitelist = origin.split(/\s*,\s*/);
            } else {
              whitelist = origin;
            }
```

or 

```javascript
            let whitelist = origin;

            if (typeof origin === 'function') {
              whitelist = await origin(ctx);
            } else if (typeof origin === 'string') {
              whitelist = origin.split(/\s*,\s*/);
            }
```

Reply:
You cannot simplify like that :) the return of the origin function must be treated too without your refactor you expect the function to return the final value.

Reply:
I see, I thought the function would return directly the array, not the string. If it can return both, then we cannot simplify yes ! I make it resolve then :)